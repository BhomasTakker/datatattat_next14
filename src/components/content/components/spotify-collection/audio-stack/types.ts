import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
