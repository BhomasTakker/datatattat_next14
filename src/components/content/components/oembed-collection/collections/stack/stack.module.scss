.root {
	.list {
		display: flex;
		flex-direction: column;
		align-items: center;
		width: 100%;
		li {
			// sm, md, etc
			width: 100%;
		}
	}
}
