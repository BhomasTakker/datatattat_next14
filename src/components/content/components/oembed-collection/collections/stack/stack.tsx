import { OEmbed } from "@/types/data-structures/oembed";
import { ClientOembed } from "../../../content-oembed/client-component";
import styles from "./stack.module.scss";

export type OembedStackProps = {};

const renderOembed = (item: OEmbed) => {
	const { src, html } = item || {};

	return (
		<li key={item.id} data-testid="content-oembed" className={styles.root}>
			{html && <ClientOembed html={html} />}
		</li>
	);
};

const renderMethod = (collection: OEmbed[] = [], props: OembedStackProps) => {
	return (
		<ul className={`${styles.list}`}>
			{collection.map((item) => renderOembed(item))}
		</ul>
	);
};

const oembedStack = {
	styles,
	renderMethod,
};

export default oembedStack;
