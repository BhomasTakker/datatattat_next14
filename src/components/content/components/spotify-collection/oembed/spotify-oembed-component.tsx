"use client";

import { useEffect, useRef, useState } from "react";
import { ClientOembed } from "../../oembed-collection/content-oembed/client-component";
import { OEmbed } from "@/types/data-structures/oembed";

// Spotify utils?
const getOembedType = (item: SpotifyOembed) => {
	const { iframe_url = "" } = item;
	let oembedType;

	switch (true) {
		case iframe_url.includes("embed/album"):
			oembedType = "album";
			break;
		case iframe_url.includes("embed/artist"):
			oembedType = "artist";
			break;
		case iframe_url.includes("embed/track"):
			oembedType = "track";
			break;
		case iframe_url.includes("embed/show"):
			oembedType = "show";
			break;
		case iframe_url.includes("embed/playlist"):
			oembedType = "playlist";
			break;
		case iframe_url.includes("embed/episode"):
			oembedType = "episode";
			break;
	}

	return oembedType;
};

type SpotifyOembed = {
	iframe_url: string;
} & OEmbed;

type SpotifyOembedComponentProps = { item: SpotifyOembed };
export const SpotifyOembedComponent = ({
	item,
}: SpotifyOembedComponentProps) => {
	const ref = useRef<HTMLDivElement>(null);
	const [size, setSize] = useState({ width: 0, height: 0 });
	const { html = "" } = item;
	let oembedType = getOembedType(item);
	console.log(`OEmbed Type: ${oembedType}`);

	const updateSize = ["album", "artist", "playlist"].includes(oembedType || "");

	useEffect(() => {
		if (!ref.current) return;

		const resizeObserver = new ResizeObserver((entries) => {
			for (let entry of entries) {
				const { width, height } = entry.contentRect;
				console.log(
					`Div size changed - Width: ${width}px, Height: ${height}px`
				);
				setSize({ width, height });
			}
		});

		resizeObserver.observe(ref.current);

		// Cleanup function to disconnect observer
		return () => {
			resizeObserver.disconnect();
		};
	}, []);

	// maybe just set height
	// apply size to html
	// available or set

	const height = Math.max(size.height, 800);
	const newHtml = html.replace(/height="\d+"/, `height="${height}"`);
	const useHtml = updateSize ? newHtml : html;

	return (
		<div ref={ref}>
			<ClientOembed html={useHtml} />
		</div>
	);
};
