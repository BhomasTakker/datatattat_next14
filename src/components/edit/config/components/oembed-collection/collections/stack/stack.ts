import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const OEMBED_STACK_CONFIG: InputListProps = {
	id: "stack",
	type: EditInputs.inputList,
	label: "Stack",
	inputs: [
		{
			id: "oembedComponentTitle",
			type: EditInputs.title,
			title: "Oembed Stack Component",
		},
	],
};
