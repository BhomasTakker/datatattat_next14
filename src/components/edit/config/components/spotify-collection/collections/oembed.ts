import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";
import { getWithConfig, QueryOptions } from "../../../query/_with-config";

export const SPOTIFY_OEMBED_CONFIG: InputListProps = {
	id: "spotifyOembed",
	type: EditInputs.inputList,
	label: "Spotify Oembed",
	inputs: [
		{
			id: "oembedComponentTitle",
			type: EditInputs.title,
			title: "Spotify Oembed Component",
		},
		getWithConfig({
			options: [
				// we need to  add this after we determine variant
				QueryOptions.SPOTIFY_OEMBED,
			],
			defaultSelection: QueryOptions.SPOTIFY_OEMBED,
		}),
	],
};
