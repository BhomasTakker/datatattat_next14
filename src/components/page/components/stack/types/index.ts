import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
