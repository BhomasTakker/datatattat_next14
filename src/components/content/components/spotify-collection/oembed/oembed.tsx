import { OEmbed } from "@/types/data-structures/oembed";
import styles from "./oembed.module.scss";
import { SpotifyOembedComponent } from "./spotify-oembed-component";

export type OembedStackProps = {};

type SpotifyOembed = {
	iframe_url: string;
} & OEmbed;

const renderMethod = (items: SpotifyOembed[] = [], props: OembedStackProps) => {
	const [item] = items;

	return <SpotifyOembedComponent item={item} />;
};

const spotifyOembed = {
	styles,
	renderMethod,
};

export default spotifyOembed;
