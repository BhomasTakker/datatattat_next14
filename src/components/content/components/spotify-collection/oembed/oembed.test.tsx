import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import spotifyOembed from "./oembed";
import { OEmbed } from "@/types/data-structures/oembed";
import { ContainerHeightOptions } from "@/components/page/components/stack/types";

// Mock the SpotifyOembedComponent
jest.mock("./spotify-oembed-component", () => ({
	SpotifyOembedComponent: ({
		item,
		height,
	}: {
		item: any;
		height: ContainerHeightOptions;
	}) => (
		<div data-testid="spotify-oembed-component">
			<div data-testid="item-id">{item?.id}</div>
			<div data-testid="item-iframe-url">{item?.iframe_url}</div>
			<div data-testid="height">{height}</div>
		</div>
	),
}));

// Mock styles
jest.mock("./oembed.module.scss", () => ({
	default: {
		container: "mocked-container-class",
		item: "mocked-item-class",
	},
}));

describe("spotifyOembed", () => {
	describe("renderMethod", () => {
		const mockOembedItem: OEmbed & { iframe_url: string } = {
			id: "test-id",
			title: "Test Track",
			type: "rich",
			version: "1.0",
			provider_name: "Spotify",
			provider_url: "https://open.spotify.com",
			html: '<iframe src="https://open.spotify.com/embed/track/123" width="300" height="380"></iframe>',
			iframe_url: "https://open.spotify.com/embed/track/123",
			width: 300,
			height: 380,
		};

		it("renders SpotifyOembedComponent with the first item", () => {
			const items = [mockOembedItem];
			const options = {};

			const result = spotifyOembed.renderMethod(items, options);
			const { container } = render(<div>{result}</div>);

			expect(
				screen.getByTestId("spotify-oembed-component")
			).toBeInTheDocument();
			expect(screen.getByTestId("item-id")).toHaveTextContent("test-id");
			expect(screen.getByTestId("item-iframe-url")).toHaveTextContent(
				"https://open.spotify.com/embed/track/123"
			);
		});

		it("uses default height when no height option is provided", () => {
			const items = [mockOembedItem];
			const options = {};

			const result = spotifyOembed.renderMethod(items, options);
			render(<div>{result}</div>);

			expect(screen.getByTestId("height")).toHaveTextContent(
				ContainerHeightOptions.MD
			);
		});

		it("uses provided height option", () => {
			const items = [mockOembedItem];
			const options = { height: ContainerHeightOptions.LG };

			const result = spotifyOembed.renderMethod(items, options);
			render(<div>{result}</div>);

			expect(screen.getByTestId("height")).toHaveTextContent(
				ContainerHeightOptions.LG
			);
		});

		it("renders with different height options", () => {
			const items = [mockOembedItem];
			const heightOptions = [
				ContainerHeightOptions.XS,
				ContainerHeightOptions.SM,
				ContainerHeightOptions.MD,
				ContainerHeightOptions.LG,
				ContainerHeightOptions.XL,
			];

			heightOptions.forEach((height) => {
				const { unmount } = render(
					<div>{spotifyOembed.renderMethod(items, { height })}</div>
				);

				expect(screen.getByTestId("height")).toHaveTextContent(height);
				unmount();
			});
		});

		it("renders with empty items array", () => {
			const items: any[] = [];
			const options = {};

			const result = spotifyOembed.renderMethod(items, options);
			render(<div>{result}</div>);

			expect(
				screen.getByTestId("spotify-oembed-component")
			).toBeInTheDocument();
			expect(screen.getByTestId("item-id")).toHaveTextContent("");
		});

		it("renders with undefined items", () => {
			const options = {};

			const result = spotifyOembed.renderMethod(undefined, options);
			render(<div>{result}</div>);

			expect(
				screen.getByTestId("spotify-oembed-component")
			).toBeInTheDocument();
			expect(screen.getByTestId("item-id")).toHaveTextContent("");
		});

		it("only uses the first item when multiple items are provided", () => {
			const secondItem: OEmbed & { iframe_url: string } = {
				...mockOembedItem,
				id: "second-id",
				title: "Second Track",
				iframe_url: "https://open.spotify.com/embed/track/456",
			};

			const items = [mockOembedItem, secondItem];
			const options = {};

			const result = spotifyOembed.renderMethod(items, options);
			render(<div>{result}</div>);

			expect(screen.getByTestId("item-id")).toHaveTextContent("test-id");
			expect(screen.getByTestId("item-id")).not.toHaveTextContent("second-id");
		});

		it("handles different Spotify URL types", () => {
			const spotifyUrls = [
				"https://open.spotify.com/embed/track/123",
				"https://open.spotify.com/embed/album/456",
				"https://open.spotify.com/embed/artist/789",
				"https://open.spotify.com/embed/playlist/abc",
				"https://open.spotify.com/embed/show/def",
				"https://open.spotify.com/embed/episode/ghi",
			];

			spotifyUrls.forEach((iframe_url, index) => {
				const item = {
					...mockOembedItem,
					id: `test-id-${index}`,
					iframe_url,
				};

				const { unmount } = render(
					<div>{spotifyOembed.renderMethod([item], {})}</div>
				);

				expect(screen.getByTestId("item-iframe-url")).toHaveTextContent(
					iframe_url
				);
				unmount();
			});
		});
	});

	describe("exports", () => {
		it("exports styles", () => {
			expect(spotifyOembed.styles).toBeDefined();
			expect(spotifyOembed.styles).toEqual({
				default: {
					container: "mocked-container-class",
					item: "mocked-item-class",
				},
			});
		});

		it("exports renderMethod", () => {
			expect(spotifyOembed.renderMethod).toBeDefined();
			expect(typeof spotifyOembed.renderMethod).toBe("function");
		});
	});

	describe("type compatibility", () => {
		it("accepts valid SpotifyOembed items", () => {
			const validItem: OEmbed & { iframe_url: string } = {
				id: "test",
				title: "Test",
				type: "rich",
				version: "1.0",
				provider_name: "Spotify",
				provider_url: "https://open.spotify.com",
				html: "<iframe></iframe>",
				iframe_url: "https://open.spotify.com/embed/track/123",
			};

			expect(() => {
				const result = spotifyOembed.renderMethod([validItem], {});
				render(<div>{result}</div>);
			}).not.toThrow();
		});

		it("handles items with additional OEmbed properties", () => {
			const itemWithExtras: OEmbed & { iframe_url: string } = {
				id: "test",
				title: "Test",
				type: "rich",
				version: "1.0",
				provider_name: "Spotify",
				provider_url: "https://open.spotify.com",
				html: "<iframe></iframe>",
				iframe_url: "https://open.spotify.com/embed/track/123",
				author_name: "Test Artist",
				author_url: "https://open.spotify.com/artist/123",
				thumbnail_url: "https://example.com/thumb.jpg",
				width: 300,
				height: 380,
				cache_age: 3600,
				customProperty: "custom value",
			};

			expect(() => {
				const result = spotifyOembed.renderMethod([itemWithExtras], {});
				render(<div>{result}</div>);
			}).not.toThrow();

			expect(screen.getByTestId("item-id")).toHaveTextContent("test");
		});
	});

	describe("Snaps", () => {
		it("renders spotifyOembed renderMethod unchanged", () => {
			const items = [
				{
					id: "test-id",
					title: "Test Track",
					type: "rich",
					version: "1.0",
					provider_name: "Spotify",
					provider_url: "https://open.spotify.com",
					html: '<iframe src="https://open.spotify.com/embed/track/123" width="300" height="380"></iframe>',
					iframe_url: "https://open.spotify.com/embed/track/123",
					width: 300,
					height: 380,
				},
			];
			const options = { height: ContainerHeightOptions.LG };

			const result = spotifyOembed.renderMethod(items, options);
			const { container } = render(<div>{result}</div>);

			expect(container).toMatchSnapshot();
		});
	});
});
