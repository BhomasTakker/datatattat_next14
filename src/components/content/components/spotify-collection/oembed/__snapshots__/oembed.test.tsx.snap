// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`spotifyOembed Snaps renders spotifyOembed renderMethod unchanged 1`] = `
<div>
  <div>
    <div
      data-testid="spotify-oembed-component"
    >
      <div
        data-testid="item-id"
      >
        test-id
      </div>
      <div
        data-testid="item-iframe-url"
      >
        https://open.spotify.com/embed/track/123
      </div>
      <div
        data-testid="height"
      >
        LG
      </div>
    </div>
  </div>
</div>
`;
